File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closeToOpen = {
            ")":"(","}":"{","]":"["
        }

        for c in s:
            if c in closeToOpen.values():
                stack.append(c)
            elif c in closeToOpen:
                if not stack or stack[-1]!=closeToOpen[c]:
                    return False
                stack.pop()
        return not stack
            

        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> match = new HashMap<>();
        match.put(')', '(');
        match.put(']', '[');
        match.put('}', '{');

        Stack<Character> stack = new Stack<>();
        // Deque<Type> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()){
            if (c == '(' || c =='[' || c=='{'){
                stack.push(c);
            }
            else{
                if (stack.isEmpty()) return false;   
                if (stack.pop()!=match.get(c)){
                    return false;
                }
                

                }
            }
        return stack.isEmpty();
        }
        


    }
